tests/test_diff_rigid.py

```diff
@@ -0,0 +1,216 @@
+            dofs_info=rigid_solver.dofs_info,
+            geoms_state=rigid_solver.geoms_state,
+            geoms_info=rigid_solver.geoms_info,
+            entities_state=rigid_solver.entities_state,
+            entities_info=rigid_solver.entities_info,
+            rigid_global_info=rigid_solver._rigid_global_info,
+            static_rigid_sim_config=rigid_solver._static_rigid_sim_config,
+            contact_island_state=constraint_solver.contact_island.contact_island_state,
+            static_rigid_sim_cache_key=rigid_solver._static_rigid_sim_cache_key,
+        )
+        rigid_solver._func_constraint_clear()
+        constraint_solver.add_equality_constraints()
+        rigid_solver.collider.detection()
+        constraint_solver.add_frictionloss_constraints()
+        constraint_solver.add_collision_constraints()
+        constraint_solver.add_joint_limit_constraints()
+        constraint_solver_resolve()
+
+    def compute_loss(input_mass, input_jac, input_aref, input_efc_D, input_force):
+        rigid_solver._rigid_global_info.mass_mat.from_numpy(input_mass.detach().cpu().numpy())
+        constraint_solver.constraint_state.jac.from_numpy(input_jac.detach().cpu().numpy())
+        constraint_solver.constraint_state.aref.from_numpy(input_aref.detach().cpu().numpy())
+        constraint_solver.constraint_state.efc_D.from_numpy(input_efc_D.detach().cpu().numpy())
+        rigid_solver.dofs_state.force.from_numpy(input_force.detach().cpu().numpy())
+
+        # Recompute acc_smooth from the updated input variables
+        updated_acc_smooth = torch.linalg.solve(input_mass.squeeze(-1), input_force.squeeze(-1))
+        input_acc_smooth = updated_acc_smooth.unsqueeze(-1).detach().cpu().numpy()
+
+        rigid_solver.dofs_state.acc_smooth.from_numpy(input_acc_smooth)
+
+        constraint_solver_resolve()
+        output_qacc = constraint_solver.qacc.to_numpy()
+        th_output_qacc = torch.from_numpy(output_qacc).to(device=target_qacc.device)
+        loss = ((th_output_qacc - target_qacc) ** 2).mean()
+        return loss
+
+    # Step once to compute constraint solver's inputs: [mass], [jac], [aref], [efc_D], [force]. We compute gradients for them.
+    scene_step()
+    init_input_mass = ti_to_torch(rigid_solver._rigid_global_info.mass_mat)
+    init_input_jac = ti_to_torch(constraint_solver.constraint_state.jac)
+    init_input_aref = ti_to_torch(constraint_solver.constraint_state.aref)
+    init_input_efc_D = ti_to_torch(constraint_solver.constraint_state.efc_D)
+    init_input_force = ti_to_torch(rigid_solver.dofs_state.force)
+
+    # [acc_smooth] is dependent on [force], not an independent input ---> No need to compute gradient for it
+    init_input_acc_smooth = ti_to_torch(rigid_solver.dofs_state.acc_smooth)
+
+    # Initial output of the constraint solver
+    init_output_qacc = ti_to_torch(constraint_solver.qacc)
+    target_qacc = torch.rand_like(init_output_qacc) * init_output_qacc.abs().mean()
```

Comment:
You should use numpy random instead of torch rand, because torch rand is machine specific and backend specific (generated numbers for the same seed are different between CPU, CUDA, MPS...etc), which is undesirable as it increases flakiness. See `test_render_api_advanced` as an example.